_posts/2020-02-09-single-responsibility-principle.md

```diff
@@ -0,0 +1,271 @@
+
+When components (not only methods and classes, but also modules and distributed services) are connected with queues (either in-memory, in-process `Queue`s, or distributed message brokers such as Kafka), we can easily monitor the sizes of the backlogs of the queues in the pipeline. Matt Welsh, the guy who proposed the [staged event-driven architecture](https://en.wikipedia.org/wiki/Staged_event-driven_architecture), regarded this observability of load and resource bottlenecks as [the most important contribution of SEDA](http://matt-welsh.blogspot.com/2010/07/retrospective-on-seda.html).
+
+Decoupled services could be scaled up and down independently in response to the changing load, without overuse of resources. Within an application, we can control the distribution of CPU resources between method, class, and module responsibilities by sizing the corresponding thread pools. `ThreadPoolExecutor` even supports dynamic reconfiguration in runtime via [`setCorePoolSize()`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ThreadPoolExecutor.html#setCorePoolSize%28int%29) method.
+
+**When microservices have focused responsibilities, it also helps to investigate incidents.** If we monitor the request success rates and health status of each service, and see that one service which connects to a particular database is failing or unavailable, we may assume that the root problem of lies in this database rather than any other part of the system.
+
+However, despite of the advantages of finer-grained monitoring and scaling, **splitting responsibilities between smaller services generally increases the burden of operating the system.** Smaller services mean more work:
+- Setting up and operating intermediate message queues (like Kafka) between the services.
+ - DevOps: setting up and managing separate delivery pipelines, monitoring, configuration, machine and container images.
+ - Deployment and orchestration: Kubernetes doesn't fully alleviate it.
+ - To ensure [rollback safety](https://aws.amazon.com/builders-library/ensuring-rollback-safety-during-deployments/), the deployments should be multi-phase, shared state and messages sent between services should be versioned.
+
+### Reliability
+Reliability is the first software quality in the list that we mostly hurt, not aid when we split smaller responsibilities between the components. Although, if engineered properly, microservices can imporve availability: when one service is sick, others might still serve something for the users, the inherent [fallability of distributed systems](https://en.wikipedia.org/wiki/Fallacies_of_distributed_computing) hits harder. Also, "if engineered properly" is an important caveat :)
```

Comment:
Why do we hurt reliability?

Reply:
Expanded and rephrased this part a bit, mentioned network partition and degradation.